### PR TITLE
:sparkles: add comments for the exported methods in generated controllers (only v3+)

### LINKED ISSUE #1606: Add comments for exported methods in generated controllers
It would be great to have the exported methods in the generate controllers documented.

i.e. 
```
// Reconcile ...
func (r *...) Reconcile(req ctrl.Request) (ctrl.Result, error) {
...

// SetupWithManager ...
func (r *...) SetupWithManager(mgr ctrl.Manager) error {
...
```

Would be happy to submit a PR.

/kind feature

### PR DESCRIPTION
**Pull request by bvwells**
**Description**
Add comments for exported methods in generated controllers in order to provide help for users. 

**Motivation**
Closes #1606.

### CODE REVIEW

**Comment by bvwells** [comment]
Need to sort out issues running make generate...

File: pkg/plugin/v3/scaffolds/internal/templates/config/controller/controller.go [hunk 1]
```diff
@@ -76,6 +76,13 @@ type {{ .Resource.Kind }}Reconciler struct {
 // +kubebuilder:rbac:groups={{ .Resource.Domain }},resources={{ .Resource.Plural }},verbs=get;list;watch;create;update;patch;delete
 // +kubebuilder:rbac:groups={{ .Resource.Domain }},resources={{ .Resource.Plural }}/status,verbs=get;update;patch
 
+// Reconcile reads that state of the cluster for a {{ .Resource.Kind }} object 
+// and makes changes based on the state and what is in the {{ .Resource.Kind }}.Spec.
+// TODO(user): Modify this Reconcile function to implement your Controller logic.
+// Note:
+// The Controller will requeue the Request to be processed again if the returned
+// error is non-nil or Result.Requeue is true, otherwise upon completion it will
+// remove the work from the queue.
```
Comment (camilamacedo86) [inline pkg/plugin/v3/scaffolds/internal/templates/config/controller/controller.go:85]: Note that we have more requeue options. So, I am afraid that we could not add this text. 

The following are a few possible return options to restart the Reconcile:

- With the error:
return ctrl.Result{}, err

- Without an error:
return ctrl.Result{Requeue: true}, nil

- Therefore, to stop the Reconcile, use:
return ctrl.Result{}, nil

- Reconcile again after X time:
return ctrl.Result{RequeueAfter: nextRun.Sub(r.Now())}, nil

Note: For more details, check the Reconcile and its [Result godoc](https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/reconcile).

File: pkg/plugin/v3/scaffolds/internal/templates/config/controller/controller.go [hunk 2]
```diff
@@ -76,6 +76,13 @@ type {{ .Resource.Kind }}Reconciler struct {
 // +kubebuilder:rbac:groups={{ .Resource.Domain }},resources={{ .Resource.Plural }},verbs=get;list;watch;create;update;patch;delete
 // +kubebuilder:rbac:groups={{ .Resource.Domain }},resources={{ .Resource.Plural }}/status,verbs=get;update;patch
 
+// Reconcile reads that state of the cluster for a {{ .Resource.Kind }} object 
+// and makes changes based on the state and what is in the {{ .Resource.Kind }}.Spec.
+// TODO(user): Modify this Reconcile function to implement your Controller logic.
```
Comment (camilamacedo86) [inline pkg/plugin/v3/scaffolds/internal/templates/config/controller/controller.go:81]: I am not sure if we can say that Reconcile read the state of the cluster. The reconcile func will be trigged in many scenarios to reconconcile the data of an object. See the docs: https://github.com/kubernetes-sigs/controller-runtime/blob/v0.6.3/pkg/reconcile/reconcile.go#L88.

**Review by camilamacedo86 — CHANGES_REQUESTED** [review]
I am all in to make the code better and more intuitive by providing an aditional info/doc. However, note that we need to be consisus with its definitions and aligned with the controller-runtime def where it is implemented and make it easier to kept maintaned. 

So, that a good option would add a very generic and small text to point users check its docs in the conroller-runtime. Then, the best approach here would we need to use the tag as well to ensure that the users would get the docs according to the controller-runtime dep used. E.g; 

> https://github.com/kubernetes-sigs/controller-runtime/blob/v0.6.3/pkg/reconcile/reconcile.go

The tag above is `v0.6.3`. Note that the controller-runtime version used by the projects is defined in: https://github.com/kubernetes-sigs/kubebuilder/blob/903abf8f4185e3e9b80708ef084eae18486e9e6e/pkg/plugin/v3/scaffolds/init.go#L40-L41. And then, to pass the var to the Bollerplate we can do as it is done for the Go.mod. See: https://github.com/kubernetes-sigs/kubebuilder/blob/903abf8f4185e3e9b80708ef084eae18486e9e6e/pkg/plugin/v3/scaffolds/init.go#L112

**Comment by bvwells** [comment]
Thanks for the review @camilamacedo86. The comment was copied across from operator-framework/operator-sdk prior it aligned with kubebuilder.

I'll see if I can reword the comment as per your suggestions.

**Comment by camilamacedo86** [comment]
Hi @bvwells,

> The comment was copied across from operator-framework/operator-sdk prior it aligned with kubebuilder.

You might be looking at some code that was describing the code/example scaffolded in the old versions.  However, in POV SDK and Kube regards the controller/reconcile should be aligned controller-runtime where its implementation came from. I am looking for your improvements. Really thank you for your contribution 🥇

**Comment by bvwells** [comment]
Hi @camilamacedo86, yes in operator-sdk there was documentation describing the reconcile loop. See here

https://github.com/operator-framework/operator-sdk/blob/53b00d125fb12515cd74fb169149913b401c8995/internal/scaffold/controller_kind.go#L208

I noticed that this documentation disappeared when operator-sdk aligned with kubebuilder in v0.0.19. This is why I submitted the issue.

**Comment by bvwells** [comment]
I'll try to sketch something out this weekend.

File: pkg/plugin/v3/scaffolds/internal/templates/config/controller/controller.go (hunk 2 above)
Comment (bvwells) [reply pkg/plugin/v3/scaffolds/internal/templates/config/controller/controller.go]: Updated the comments. Let me know what you think...

File: pkg/plugin/v3/scaffolds/internal/templates/config/controller/controller.go (hunk 1 above)
Comment (bvwells) [reply pkg/plugin/v3/scaffolds/internal/templates/config/controller/controller.go]: Added documentation pointing to the controller runtime.

**Comment by bvwells** [comment]
Hi @camilamacedo86, hope you had a good weekend! Made a few changes to the Reconcile documentation. Let me know what you think. I'm more than happy to iterate on it.

**Review by camilamacedo86 — APPROVED** [review]
Really thank you for your contribution. 
It shows great for me. 👍 Could you just rebase it with the master and re-gen the testdata with `make generate` to update this PR? 
 
c/c @gab-satchi @estroz WDYT?

**Comment by camilamacedo86** [comment]
/approve

However, note it will not pass in the CI before you update it with the master branch.

**Comment by bvwells** [comment]
Thanks for the reviews @camilamacedo86! All rebased to master.

**Review by Adirio — APPROVED** [review]
LGTM
I'm not setting the flag to let @camilamacedo86 have the last word as she already set the approved flag and if I set the lgtm one it will get merged.

**Comment by camilamacedo86** [comment]
> I'm not setting the lgtm flag to let @camilamacedo86 have the last word as she already set the approved flag and if I set the lgtm one it will get merged.

Hi @Adirio, 

Thank you for your review on that. It shows fine for me as well. I think it might be helpful to end-users in order to address common questions. So, 

/lgtm

**Comment by bvwells** [comment]
Thanks @camilamacedo86 and @Adirio for your help getting this over the line!

**Pull request merged**